litmus-portal/graphql-server/pkg/k8s/cluster.go

```diff
@@ -71,50 +74,150 @@ func ClusterResource(manifest string, namespace string) (*unstructured.Unstructu
+/*
+	This function returns the endpoint of the server by which external agents can communicate.
+	The order of generating the endpoint is based on different network type:
+	- Ingress
+	- LoadBalancer > NodePort > ClusterIP
+*/
+func GetServerEndpoint() (string, error) {
 	var (
-		Nodeport       int32
-		ExternalIP     string
-		InternalIP     string
-		LitmusPortalNS = os.Getenv("LITMUS_PORTAL_NAMESPACE")
+		NodePort          int32
+		Port              int32
+		InternalIP        string
+		IngressPath       string
+		IPAddress         string
+		Scheme            string
+		FinalUrl          string
+		ServerServiceName = os.Getenv("SERVER_SERVICE_NAME")
+		NodeName          = os.Getenv("NODE_NAME")
+		LitmusPortalNS    = os.Getenv("LITMUS_PORTAL_NAMESPACE")
+		Ingress           = os.Getenv("INGRESS")
+		IngressName       = os.Getenv("INGRESS_NAME")
 	)
 
 	clientset, err := GetGenericK8sClient()
 	if err != nil {
 		return "", err
 	}
 
-	podList, _ := clientset.CoreV1().Pods(LitmusPortalNS).List(metaV1.ListOptions{
-		LabelSelector: "component=litmusportal-server",
-	})
+	if Ingress == "true" {
+		getIng, err := clientset.ExtensionsV1beta1().Ingresses(LitmusPortalNS).Get(IngressName, metaV1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
 
-	svc, err := clientset.CoreV1().Services(LitmusPortalNS).Get("litmusportal-server-service", metaV1.GetOptions{})
-	if err != nil {
-		return "", err
-	}
+		/*
+			Priorities of retrieving Ingress endpoint
+			1. hostname
+			2. IPAddress
+		*/
+
+		if len(getIng.Status.LoadBalancer.Ingress) > 0 {
+			if len(getIng.Spec.Rules) > 0 {
+				if getIng.Spec.Rules[0].Host != "" {
+					IPAddress = getIng.Spec.Rules[0].Host
+				} else if getIng.Status.LoadBalancer.Ingress[0].IP != "" {
+					IPAddress = getIng.Status.LoadBalancer.Ingress[0].IP
+				}
+			} else {
+				return "", errors.New("Ingress rules are not present")
+			}
+		} else {
+			return "", errors.New("IP Address or HostName not generated")
+		}
 
-	for _, port := range svc.Spec.Ports {
-		if port.Name == "graphql-server" {
-			Nodeport = port.NodePort
+		if IPAddress == "" {
+			return "", errors.New("IP Address or Hostname is not available in the ingress of " + IngressName)
 		}
-	}
 
-	nodeIP, err := clientset.CoreV1().Nodes().Get(podList.Items[0].Spec.NodeName, metaV1.GetOptions{})
-	if err != nil {
-		return "", err
-	}
+		for _, rule := range getIng.Spec.Rules {
+			for _, path := range rule.HTTP.Paths {
+				if path.Backend.ServiceName == ServerServiceName {
+					log.Print(path.Path)
+					path_arr := strings.Split(path.Path, "/")
+					if path_arr[len(path_arr)-1] == "(.*)" {
+						path_arr[len(path_arr)-1] = "query"
+					} else {
+						path_arr = append(path_arr, "query")
+					}
+
+					log.Print(path_arr)
+					for el, p := range path_arr {
+						if p != "" {
+							if el == len(path_arr)-1 {
+								IngressPath += p
+							} else {
+								IngressPath += p + "/"
+							}
+						}
+					}
```

Comment:
```suggestion
					IngressPath = strings.Join(path_arr[:], "/")
```

Reply:
It's good, but I need to check some edge case with strings.Join.

Reply:
sure, for the empty strings i have already suggested a change so it gets checked while generating the array itself.